### DISCUSSION TITLE
Phoenix (elixir) compatibility.

### DISCUSSION THREAD

**Discussion by karloluis** [chrispahm/chartjs-plugin-dragdata#137] [Q&A]
Has anyone successfully integrated this plugin in Phoenix with Liveview?

I'm using a radar chart on the main ChartJS and want to capture the move events for the nodes (dots) but they won't move. Using the `phx-click` binding doesn't really help much.

**Comment by karloluis** [comment]
I tried intercepting the minified code and found that the call for `(function (global, factory){})(this, () => {...})` the `this` is showing up as empty object rather than `Window` compared to my POC with the cdn link and vanilla JS. 

Not sure why this happens. I suspect it might be how I'm importing the dependency from `node_modules`.

**Comment by artus9033** [comment]
Hi @karloluis, it would be great if you could provide the code you are using or even a MRE repo, since now I can only guess.

I believe you might be doing something like `import ChartJSDragDataPlugin from "../node_modules/chartjs-plugin-dragdata/dist/chartjs-plugin-dragdata";`, which is invalid as you are importing a UMD module that is prepared to work in a browser environment, from your code that is processed by a bundler and most likely allows you to use ES modules.

Instead, `import "chartjs-plugin-dragdata";` should work in Phoenix (at least works for me if I install with npm to `assets/`). That way, you import the package and `package.json` defines proper exports for each of the import styles that the bundler expects. In this case, it would use the `"import"` rule and use `dist/chartjs-plugin-dragdata.esm.js` which is an ES entrypoint.

I event tested it, and it works properly ;)
<img width="994" alt="image" src="https://github.com/user-attachments/assets/6c8ef3fb-2d67-44d4-a38c-fd11f7e3bb3b">

**Comment by karloluis** [comment]
I hopped in half asleep to copy over some code, seems I'm still missing something. The code is in `/assets` with `/assets/node_modules` and `assets/js/app.js.
_I'm hoping to have some time to prep a mock repo over the weekend._ Maybe you can spot it from this code sample?


proj/live/proj_web/live/page_live/index.ex
```elixir
@impl true
  def render(assigns) do
    ~H"""
      <div>
        <canvas id="chart" phx-hook="radarChart"></canvas>
      </div>
    """
  end
```

./assets/js/app.js
```javascript
import "phoenix_html"
import { Socket } from "phoenix"
import { LiveSocket } from "phoenix_live_view"
import topbar from "../vendor/topbar"
import Chart from "chart.js/auto"
import "chartjs-plugin-dragdata"

hooks.radarChart = {
  dataset() {
    // TODO
  },
  mounted() {
    const dataElement = this.el
    const ctx = document.getElementById(dataElement.id)

    new Chart(ctx, dataOptions)
  },
  ....
}

// Usage further down
let liveSocket = new LiveSocket("/live", Socket, {
  hooks,
  longPollFallbackMs: 2500,
  params: { _csrf_token: csrfToken },
  metadata: {
    // click: (e, _el) => {
    //   console.log(e)
    //   return { offsetX: e.offsetX, offsetY: e.offsetY }
    // },
  },
})
```

https://github.com/user-attachments/assets/76a334ff-70f4-4518-a6a7-2d836740e314

**Comment by karloluis** [comment]
<img width="495" alt="Screenshot 2024-08-22 at 1 22 36 AM" src="https://github.com/user-attachments/assets/d84d75fc-d09a-41c7-8329-f2b49aeae17a">
<img width="1016" alt="Screenshot 2024-08-22 at 1 22 42 AM" src="https://github.com/user-attachments/assets/2958223f-9c36-4457-b8a3-62854f6a9acf">
